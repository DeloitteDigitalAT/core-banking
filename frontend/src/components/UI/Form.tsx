import React, { useState, useEffect } from "react";
import Input from "./Input";
import Button from "./Button";

interface CommonType {
  id: number;
  type: string;
}
interface InputType extends CommonType {
  label: string;
  placeholder: string;
  value: string;
  required: boolean;
  readonly: boolean;
  error: boolean;
  divClass: string;
  text?: never;
  handler?: never;
  active?: never;
}

interface ButtonType extends CommonType {
  text: string;
  handler: () => void;
  active?: boolean;
  divClass: string;
  required?: never;
  error?: never;
  value?: never;
  label?: never;
  placeholder?: never;
  readonly?: never;
}

type ElementType = InputType | ButtonType;
interface FormType {
  elements: Array<ElementType>;
  setForm: (el: Array<ElementType>) => void;
}

const Form: React.FC<FormType> = properties => {
  const { elements, setForm } = properties;
  const [formError, setFormError] = useState("");
  useEffect(() => {
    setFormError("");
  }, []);
  const checkErrors = () => {
    const updElements = [...elements].map(element => {
      if (element.type === "input" && element.required && !element.value) {
        element.error = true;
      }
      return element;
    });
    if (!elements.find(e => e.error)) {
      setFormError("");
    }
    setForm(updElements);
  };
  const changeForm = (id: number, newValue: string) => {
    const updElements = [...elements].map(element => {
      if (element.id === id && element.type === "input") {
        if (element.error && newValue.length) {
          element.error = false;
        }
        if (element.error && !newValue.length) {
          element.error = true;
        }
        element.value = newValue;
      }
      return element;
    });
    setForm(updElements);
  };
  const clearForm = () => {
    const updElements = [...elements].map(element => {
      if (element.type === "input" && element.value) {
        element.value = "";
        element.error = false;
      }
      return element;
    });
    setForm(updElements);
  };

  return (
    <div className="grid grid-cols-6">
      {elements.map(element => {
        if (element.type === "input") {
          return (
            <div key={element.id} className={element.divClass}>
              <Input
                label={element.label!}
                placeholder={element.placeholder!}
                value={element.value!}
                required={element.required!}
                error={element.error!}
                readonly={element.readonly!}
                setInput={value => {
                  changeForm(element.id, value);
                }}
              />
            </div>
          );
        }
        if (element.type === "button") {
          return (
            <div key={element.id} className={element.divClass}>
              <Button
                text={element.text!}
                active={!elements.find(e => e.error)}
                handler={() => {
                  if (elements.find(e => e.required && !e.value)) {
                    checkErrors();
                    setFormError("Required fields must be filled in");
                  } else {
                    setFormError("");
                    element.handler!();
                    clearForm();
                  }
                }}
              />
            </div>
          );
        }
        return null;
      })}
      {elements.find(e => e.error) && <p className="whitespace-nowrap	text-mango-600 text-sm">{formError}</p>}
    </div>
  );
};

export default Form;
