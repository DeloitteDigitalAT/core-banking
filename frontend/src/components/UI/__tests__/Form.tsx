import React from "react";
import renderer from "react-test-renderer";
import Form from "../Form";

const mockForm = [
  {
    id: 1,
    type: "input",
    label: "Credit amount requested",
    placeholder: "",
    value: "123",
    required: false,
    readonly: true,
    error: false,
    divClass: "col-span-3 md:col-span-2",
  },
  {
    id: 2,
    type: "input",
    label: "Interest rate",
    placeholder: "",
    value: "456",
    required: false,
    readonly: true,
    error: false,
    divClass: "col-span-3 md:col-span-1 pl-4",
  },
  {
    id: 3,
    type: "input",
    label: "Planned emi",
    placeholder: "",
    value: "789",
    required: false,
    readonly: true,
    error: false,
    divClass: "col-span-3 md:col-span-1 md:pl-4",
  },
  {
    id: 4,
    type: "input",
    label: "Duration in months",
    placeholder: "",
    value: "24",
    required: false,
    readonly: true,
    error: false,
    divClass: "col-span-3 md:col-span-2 pl-4",
  },
  {
    id: 5,
    type: "button",
    text: "Search",
    handler: () => {},
    divClass: "col-span-6",
  },
];

describe("UI/Form component", () => {
  test("snapshot renders", () => {
    const tree = renderer.create(<Form elements={mockForm} setForm={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
