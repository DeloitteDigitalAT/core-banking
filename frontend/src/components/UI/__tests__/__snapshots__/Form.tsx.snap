// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/Form component snapshot renders 1`] = `
<div
  className="grid grid-cols-6"
>
  <div
    className="col-span-3 md:col-span-2"
  >
    <label
      className="block mt-4"
    >
      <span
        className="block uppercase tracking-wide text-xs font-semibold mb-2 text-gray-700"
      >
        Credit amount requested
      </span>
      <input
        className="form-input mt-1 block w-full text-sm border rounded py-2 px-2  outline-none"
        onChange={[Function]}
        placeholder=""
        readOnly={true}
        value="123"
      />
    </label>
  </div>
  <div
    className="col-span-3 md:col-span-1 pl-4"
  >
    <label
      className="block mt-4"
    >
      <span
        className="block uppercase tracking-wide text-xs font-semibold mb-2 text-gray-700"
      >
        Interest rate
      </span>
      <input
        className="form-input mt-1 block w-full text-sm border rounded py-2 px-2  outline-none"
        onChange={[Function]}
        placeholder=""
        readOnly={true}
        value="456"
      />
    </label>
  </div>
  <div
    className="col-span-3 md:col-span-1 md:pl-4"
  >
    <label
      className="block mt-4"
    >
      <span
        className="block uppercase tracking-wide text-xs font-semibold mb-2 text-gray-700"
      >
        Planned emi
      </span>
      <input
        className="form-input mt-1 block w-full text-sm border rounded py-2 px-2  outline-none"
        onChange={[Function]}
        placeholder=""
        readOnly={true}
        value="789"
      />
    </label>
  </div>
  <div
    className="col-span-3 md:col-span-2 pl-4"
  >
    <label
      className="block mt-4"
    >
      <span
        className="block uppercase tracking-wide text-xs font-semibold mb-2 text-gray-700"
      >
        Duration in months
      </span>
      <input
        className="form-input mt-1 block w-full text-sm border rounded py-2 px-2  outline-none"
        onChange={[Function]}
        placeholder=""
        readOnly={true}
        value="24"
      />
    </label>
  </div>
  <div
    className="col-span-6"
  >
    <button
      className="bg-gradient-to-b from-green-400 via-green-500 to-green-600 border-green-500 hover:from-green-600 hover:via-green-700 hover:to-green-800 text-sm border text-white rounded-md px-6 py-2 my-4 transition duration-500 ease select-none focus:outline-none focus:shadow-outline"
      onClick={[Function]}
      type="button"
    >
      Search
    </button>
  </div>
</div>
`;
